### PR TITLE
Draft: Php 8.1 support

### PR DESCRIPTION
**Pull request by yurichechulin**
### Added

- Support PHP `8.1`

### Changed

- Composer `2.x` is supported now

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I wrote unit tests for my code _(if tests is required for my changes)_
- [x] I have made changes in `CHANGELOG.md` file

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -34,8 +30,8 @@ jobs: # Docs: <https://git.io/JvxXE>
         uses: shivammathur/setup-php@v2 # Action page: <https://github.com/shivammathur/setup-php>
         with:
           php-version: ${{ matrix.php }}
-          extensions: mbstring
-          tools: composer:v1 # TODO update composer up to v2
+          extensions: mbstring, xdebug
```
Comment (gomzyakov) [inline .github/workflows/tests.yml:33]: Is `xdebug` necessary?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -20,11 +20,7 @@ jobs: # Docs: <https://git.io/JvxXE>
       matrix:
         setup: ['basic', 'lowest']
         illuminate: [default]
-        php: ['7.3', '7.4']
-        include:
-          - php: '7.3'
-            setup: basic
-            illuminate: '~7.0'
+        php: ['8.0', '8.1']
```
Comment (gomzyakov) [inline .github/workflows/tests.yml:23]: Why PHP 7.4 lost support?

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -48,27 +44,30 @@ jobs: # Docs: <https://git.io/JvxXE>
...
         run: composer info | grep -e phpunit -e phpstan -e illuminate -e avtocod
 
       - name: Execute tests
         run: composer test-cover
 
+      - name: Execute tests with code coverage
+        if: matrix.coverage == 'yes'
+        env:
+          XDEBUG_MODE: coverage
+        run: composer test-cover
```
Comment (gomzyakov) [inline .github/workflows/tests.yml:69]: Maybe, are we out of scope?

File: composer.json [hunk 4]
```diff
@@ -16,16 +16,16 @@
         }
     ],
     "require": {
-        "php": "^7.3",
+        "php": "^8.0",
```
Comment (gomzyakov) [inline composer.json:19]: Why not `"php": "^7.3 || ^8.0"`?

Look at https://github.com/avtocod/specs-php/pull/9

File: composer.json [hunk 5]
```diff
@@ -16,16 +16,16 @@
...
         "ext-json": "*",
         "ext-mbstring": "*",
-        "avtocod/specs": "~3.70",
-        "illuminate/support": "~6.0 || ~7.0 || ~8.0",
-        "ocramius/package-versions": "^1.4"
+        "avtocod/specs": "~3.99",
```
Comment (gomzyakov) [inline composer.json:22]: Why not `^3.0`?

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,17 @@ All notable changes to this project will be documented in this file.
...
+## UNRELEASED
+
+### Added
+
+- Minimal required PHP version now is `8.0`
+- Minimal required dependency `illuminate/support` now is `~9.0`
+
+### Changed
+
+- Composer `2.x` is supported now
```
Comment (gomzyakov) [inline CHANGELOG.md:16]: What about `xdebug`?

File: src/Structures/Source.php [hunk 7]
```diff
@@ -66,12 +66,14 @@ protected function configure($raw_data)
             foreach ($raw_data as $key => $value) {
                 switch ($key) {
                     case 'name':
+                        /** @var string|null $value */
```
Comment (gomzyakov) [inline src/Structures/Source.php:69]: Can we go without upgrading the version of `PHPStan`?

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,17 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog][keepachangelog] and this project adheres to [Semantic Versioning][semver].
 
+## UNRELEASED
+
+### Added
+
+- Minimal required PHP version now is `8.0`
+- Minimal required dependency `illuminate/support` now is `~9.0`
+
```
Comment (yurichechulin) [inline CHANGELOG.md:13]: Laravel 9 requires typed Collection<TKey, TValue> in their signature, Laravel 8 doesn't.

Phpstan throws error in that cases: 

- If we remove Collection key and value typing, it doesn't match with Laravel 9 signature.
- If we set Collection<TKey, TValue>, it doesn't match with Laravel 8 signature.


That's why we boosted minimal required Laravel version to 9.0

**Comment by Zaboday** [comment]
Resolved at https://github.com/avtocod/specs-php/releases/tag/v1.6.0

**Pull request closed without merging**